ts/session/protocols/SessionProtocol.ts

```diff
@@ -1,56 +1,303 @@
+    }
+  }
+
+  /**
+   * Called when a session is establish so we store on database this info.
+   */
+  public static async onSessionEstablished(pubkey: PubKey) {
+    // remove our existing sent timestamp for that device
+    return SessionProtocol.updateSentSessionTimestamp(pubkey.key, undefined);
+  }
+
+  public static async shouldProcessSessionRequest(
+    pubkey: PubKey,
+    messageTimestamp: number
+  ): Promise<boolean> {
+    const existingSentTimestamp =
+      (await SessionProtocol.getSentSessionRequest(pubkey.key)) || 0;
+    const existingProcessedTimestamp =
+      (await SessionProtocol.getProcessedSessionRequest(pubkey.key)) || 0;
+
+    return (
+      messageTimestamp > existingSentTimestamp &&
+      messageTimestamp > existingProcessedTimestamp
+    );
+  }
+
+  public static async onSessionRequestProcessed(pubkey: PubKey) {
+    return SessionProtocol.updateProcessedSessionTimestamp(pubkey.key, Date.now());
+  }
 
-export function sessionRequestProcessed(device: string) {
-  // TODO: this is called when we process the session request
-  // This should store the processed timestamp
-  // Again naming is crap so maybe some other name is better
+  public static reset() {
+    SessionProtocol.dbLoaded = false;
+    SessionProtocol.sentSessionsTimestamp = {};
+    SessionProtocol.processedSessionsTimestamp = {};
+  }
+
+  /**
+   * We only need to fetch once from the database, because we are the only one writing to it
+   */
+  private static async fetchFromDBIfNeeded(): Promise<void> {
+    if (!SessionProtocol.dbLoaded) {
+      const sentItem = await getItemById('sentSessionsTimestamp');
+      if (sentItem) {
+        SessionProtocol.sentSessionsTimestamp = sentItem.value;
+      } else {
+        SessionProtocol.sentSessionsTimestamp = {};
+      }
+
+      const processedItem = await getItemById('processedSessionsTimestamp');
+      if (processedItem) {
+        SessionProtocol.processedSessionsTimestamp = processedItem.value;
+      } else {
+        SessionProtocol.processedSessionsTimestamp = {};
+      }
+      SessionProtocol.dbLoaded = true;
+    }
+  }
+
+  private static async writeToDBSentSessions(): Promise<void> {
+    const data = {
+      id: 'sentSessionsTimestamp',
+      value: JSON.stringify(SessionProtocol.sentSessionsTimestamp),
+    };
+
+    await createOrUpdateItem(data);
+  }
+
+  private static async writeToDBProcessedSessions(): Promise<void> {
+    const data = {
+      id: 'processedSessionsTimestamp',
+      value: JSON.stringify(SessionProtocol.processedSessionsTimestamp),
+    };
+
+    await createOrUpdateItem(data);
+  }
+
+  /**
+   * This is a utility function to avoid duplicated code of updateSentSessionTimestamp and updateProcessedSessionTimestamp
+   */
+  private static async updateSessionTimestamp(
+    device: string,
+    timestamp: number | undefined,
+    map: StringToNumberMap
+  ): Promise<boolean> {
+    if (!timestamp) {
+      if (!!map[device]) {
```

Comment:
```suggestion
      if (device in map) {
```
this might make more sense when reading this section